### ISSUE TITLE
SQL Server 2008 pagination skip and take

### ISSUE THREAD

**Issue by devanarch** [fsprojects/SQLProvider#504]
### Description

Provider uses OFFSET FETCH clause for handling SKIP/TAKE, and as stated in code comments "this only works in >=SQL2012"

### Repro steps

1. Table SQL
```
CREATE TABLE [NU-SOFT].[Fixture](
	[Id] [int] IDENTITY(1,1) NOT NULL,
	[Name] [nvarchar](20) NOT NULL)
```
2. Query
```
query {
    for f in ctx.NuSoft.Fixture do
    where (f.Name =% "%a")
    sortBy f.Name
    skip 2
    take 3
}
```

### Expected behavior

Expected SQL for 2008 would be:
```
WITH CTE AS ( 
	SELECT 
		[f].[Id] as 'Id',
		[f].[Name] as 'Name' , 
		ROW_NUMBER() OVER(ORDER BY  [f].[Name]  ) AS RN  
	FROM [NU-SOFT].[Fixture] as [f] 
	WHERE (([f].[Name] LIKE @param1))
)
SELECT 
	[f].[Id] as 'Id',
	[f].[Name] as 'Name' 
FROM CTE [f] 
WHERE RN BETWEEN 3 AND 5
```

### Actual behavior

Actual SQL generated by provider:
```
SELECT 
	[f].[Id] as 'Id',
	[f].[Name] as 'Name' 
FROM [NU-SOFT].[Fixture] as [f] 
WHERE (([f].[Name] LIKE @param1))
ORDER BY [f].[Name] 
OFFSET 2 ROWS FETCH NEXT 3 ROWS ONLY
```

### Related information 

SQL Server 2008

**Comment by Thorium** [comment]
Hmm. Quite an awful expected SQL clause...

I think this could be done by detecting SQL-server version, e.g.
```sql
SELECT
  CASE 
     WHEN CONVERT(VARCHAR(128), SERVERPROPERTY ('productversion')) like '8%' THEN 'SQL2000'
     WHEN CONVERT(VARCHAR(128), SERVERPROPERTY ('productversion')) like '9%' THEN 'SQL2005'
     WHEN CONVERT(VARCHAR(128), SERVERPROPERTY ('productversion')) like '10.0%' THEN 'SQL2008'
     WHEN CONVERT(VARCHAR(128), SERVERPROPERTY ('productversion')) like '10.5%' THEN 'SQL2008 R2'
     WHEN CONVERT(VARCHAR(128), SERVERPROPERTY ('productversion')) like '11%' THEN 'SQL2012'
     WHEN CONVERT(VARCHAR(128), SERVERPROPERTY ('productversion')) like '12%' THEN 'SQL2014'
     WHEN CONVERT(VARCHAR(128), SERVERPROPERTY ('productversion')) like '13%' THEN 'SQL2016'     
     WHEN CONVERT(VARCHAR(128), SERVERPROPERTY ('productversion')) like '14%' THEN 'SQL2017' 
     ELSE 'FUTURE'
  END AS MajorVersion
```

when creating a connection. But is it worth of performing extra call for everyone just to support non-supported-paging of 10 years old SQL-server-version? Microsoft still has ended the basic product support for SQL Server 2008 but has some extended support left. So do we still have active users a lot for this, or are you going to update the server version anyway soon?

Is the best current workaround to run just `take (pagesize*pageIndex+pagesize)` and do the skip on client side?

**Comment by devanarch** [comment]
Yeah, i know how awful it looks, though it's the only one i can think of actually doing paging on sql side.
So far i have a modified fork (not commited though) up and running on production. 
I added boolean constructor parameter for MSSqlServerProvider (no extra call) and handled it in GenerateQueryText.

Client side skip can be quite expensive depending on how distant page user would choose to view.

I wish the decision to upgrade was mine, unfortunately it involves corporate decision making on customer side, and as such i lost all hope for that ;)

**Comment by Thorium** [comment]
Ok, running on production with custom fork is not maintainable solution. The SQLProvider code base is also moving quite fast forward.

Consider creating a PR and I'll accept that. But I would like the parameter name to be something more generic than a provider specific. Like a "CompatibilityMode" or something.

**Comment by Thorium** [comment]
Actually there is already this:

```fsharp
    let isSQL2012Orlater (con:IDbConnection) =
        try
            let reader = executeSql "SELECT SERVERPROPERTY('productversion')" con
            let version = reader.GetSqlString(0)
            match version.Value.[0..1] |> Double.TryParse with
            | true, v -> v >= 11.0
            | _ -> false
        with _ -> false
```

**Comment by devanarch** [comment]
well i noticed that one, but using that would generate the 

>  extra call for everyone just to support non-supported-paging of 10 years old SQL-server-version.

 I was wondering whether to use or not, but my thinking is (i may have missed something though):

1. at the moment of constructing `MSSqlServerProvider` connection is not available, and as such we cannot call `isSQL2012Orlater` and pass result as constructor parameter
2. information regarding version is required in `GenerateQueryText`, where connection is not passed. We would need to pass connection or result of `isSQL2012Orlater` to  `GenerateQueryText` and as such affect every provider with extra method parameter (useless for non SQL Server providers)

regarding the ServerVersion, even if method would be changed to 
```
    let isSQL2012Orlater (con:IDbConnection) =
        try
            let con = (con :?> SqlConnection)
            let version = con.ServerVersion 
            match version.[0..1] |> Double.TryParse with
            | true, v -> v >= 11.0
            | _ -> false
        with _ -> false
```

the above 2 points still stand (again, I may have missed something)

**Comment by devanarch** [comment]
Changes I made are:
[additional parameter for the type provider `MSSQLPagingCompatibility`](https://github.com/devanarch/SQLProvider/commit/b2fafffe9e9c22dac789d4ea665bfd16d6f71b53#diff-8d7a61e9b41f3e400217abeddcfb6083R55) passed as a [constructor parameter to `MSSqlServerProvider`](https://github.com/devanarch/SQLProvider/commit/b2fafffe9e9c22dac789d4ea665bfd16d6f71b53#diff-8d7a61e9b41f3e400217abeddcfb6083R55) and handled it in the `GenerateQueryText` [1](GenerateQueryText) [2](https://github.com/devanarch/SQLProvider/commit/b2fafffe9e9c22dac789d4ea665bfd16d6f71b53#diff-63b4f917ef512aaea51615d1da1b2c98R933) [3](https://github.com/devanarch/SQLProvider/commit/b2fafffe9e9c22dac789d4ea665bfd16d6f71b53#diff-63b4f917ef512aaea51615d1da1b2c98R957)

generated query uses CTE when compatibility mode is set to `RowNumber` and `SKIP` clause is present

@Thorium I know you wanted parameter to be more generic, though as for now it's specific to SQL Server, similarly to `SQLiteLibrary` being specific to SQLite

let me know what you think

**Comment by Thorium** [comment]
We have also another issue related to SQL server version, #521

**Comment by piaste** [comment]
> information regarding version is required in GenerateQueryText, where connection is not passed. We would need to pass connection or result of isSQL2012Orlater to GenerateQueryText and as such affect every provider with extra method parameter (useless for non SQL Server providers)

Good point. Fortunately there is only [one invocation of `GenerateQueryText`](https://github.com/fsprojects/SQLProvider/blob/40eef746065dff4f0778cbb8193b7f2bb2b05d1e/src/SQLProvider/SqlRuntime.QueryExpression.fs#L698), and at that point there is a connection object in scope.

Note also that the `GetColumns` method a couple lines above already makes use of this object, and since the columns are one of the arguments to `GenerateQueryText`, there is already an indirect dependency on the connection.

Given that  `GenerateQueryText` is an _internal_ interface, I think adding the connection parameter is the correct solution. Plus it's not unlikely that other providers may have a use for it - for example, both `NpgsqlConnection` and `SQLiteConnection` expose similar version properties, which could be useful for query generation purposes.

**Comment by Thorium** [comment]
GenerateQueryText is called per query. GetColumns is called only once and cached. So maybe an ugly mutable option-variable could be considered: that is very highly unlikely that the server version would ever change on-the-fly.

**Comment by piaste** [comment]
[Like this?](https://github.com/piaste/SQLProvider/blob/8e623cebd5442fa1fbcc69c4ef12e896857aa525/src/SQLProvider/Providers.MsSqlServer.fs#L533) I used a ConcurrentBag over a mutable option-variable, because the latter aren't thread safe for anything but small primitives, and since the other lookups use concurrent collections it seems thread safety was a concern.

I don't have a SQL2008 instance at hand to test it, though, but if @devanarch can confirm that it works he can submit a PR (also he did the most work with the ROWNUMBER implementatino) - with this implementation, only `Providers.MsSqlServer.fs` needs to be changed.

**Comment by Thorium** [comment]
@piaste, what is the status of this, are you going to make a PR? :-)

**Comment by piaste** [comment]
@thorium 

If @devanarch isn't interested, I guess I will.

AppVeyor offers SQL2008R2 as a service, I'll see if I can use that to test the feature.

**Comment by Thorium** [comment]
Just noticed that the latest fails for call ServerVersion on:

```
            if not (mssqlVersionCache.ContainsKey con.ConnectionString) then      
                if con.State <> ConnectionState.Open then con.Open()
                let success, version = (con :?> SqlConnection).ServerVersion |> Version.TryParse
                if success then mssqlVersionCache.TryAdd(con.ConnectionString, version) |> ignore
```

I think the concurrent dictionary should be used as get-or-add and ContainsKey.

**Comment by Thorium** [comment]
Fixed in 1.1.51
